hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/IterableStorageManager.java

```diff
@@ -51,29 +61,199 @@ public IterableStorageManager() {
      * slots used per contract via
      * {@link HandleHederaOperations#updateStorageMetadata(long, Bytes, int)}.
      *
-     * @param hederaOperations the scope of the current transaction
+     * @param enhancement the enhancement for the current transaction
      * @param allAccesses the pending changes to storage values
      * @param allSizeChanges the pending changes to storage sizes
      * @param store the writable state store
      */
     public void persistChanges(
-            @NonNull final HederaOperations hederaOperations,
+            @NonNull final Enhancement enhancement,
             @NonNull final List<StorageAccesses> allAccesses,
             @NonNull final List<StorageSizeChange> allSizeChanges,
             @NonNull final ContractStateStore store) {
-        // TODO - include storage linked list management before performance testing
+        // map to store the first storage key for each contract
+        final Map<Long, Bytes> firstKeys = new HashMap<>();
 
-        // Remove all zeroed-out slots from the linked lists
+        // Adjust the storage linked lists for each contract
         allAccesses.forEach(contractAccesses -> contractAccesses.accesses().forEach(access -> {
-            if (access.isRemoval()) {
-                store.removeSlot(new SlotKey(contractAccesses.contractNumber(), tuweniToPbjBytes(access.key())));
+            if (access.isUpdate()) {
+                var firstContractKey = contractFirstKeyOf(enhancement, contractAccesses.contractNumber());
+
+                switch (StorageAccessType.getAccessType(access)) {
+                    case REMOVAL -> firstContractKey = removeAccessedValue(
+                            store, firstContractKey, contractAccesses.contractNumber(), tuweniToPbjBytes(access.key()));
+                    case INSERTION -> firstContractKey = insertAccessedValue(
+                            store,
+                            firstContractKey,
+                            tuweniToPbjBytes(access.writtenValue()),
+                            contractAccesses.contractNumber(),
+                            tuweniToPbjBytes(access.key()));
+                    case UPDATE -> updateAccessedValue(
```

Comment:
The `SavepointStackImpl` will already be committing the modified slot values as part of committing the storage `WritableKVState` [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/WritableContractStateStore.java#L35) (note we used its pending modifications to construct the `List<StorageAccesses>` [in the first place](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/DispatchingEvmFrameState.java#L134)).

So afaict this branch is doing redundant work, given it doesn't change `prev`/`next` pointers.

Reply:
makes sense.  removed the update case.